ADBDEV-5400: Close COPY TO/FROM PROGRAM descriptors in case of error

RekGRpth opened the pull request:
Close COPY TO/FROM PROGRAM descriptors in case of error

Commit 110b825 added an ifThrow argument to the close_program_pipes function.
That commit contained a call to close_program_pipes(cstate, false) on errors.
But then this function call was lost. And now, in case of errors, open pipes are
not closed until the end of the session.

This patch adds a callback for closing handles on errors, which is called before
deleting the memory context of the copy state.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/70037

BenderArenadata commented:
Failed job Build for ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1346944

BenderArenadata commented:
Failed job Build for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1346943

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/70038

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/70039

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/70042

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/70043

RekGRpth committed:
Close COPY TO/FROM PROGRAM descriptors in case of error

Commit 110b825 added an ifThrow argument to the close_program_pipes function.
That commit contained a call to close_program_pipes(cstate, false) on errors.
But then this function call was lost. And now, in case of errors, open pipes are
not closed until the end of the session.

This patch adds a callback for closing handles on errors, which is called before
deleting the memory context of the copy state.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/70054

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1348347

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1348348

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1348345

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1348340

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1348341

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1348342

RekGRpth committed:
stabilize tests

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/70064

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1348906

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1348907

Stolb27 commented on src/backend/commands/copy.c:
```diff
@@ -7955,9 +7955,15 @@ GetTargetSeg(GpDistributionData *distData, Datum *baseValues, bool *baseNulls)
 
+static void close_program_pipes_on_reset(void *arg) {
+	CopyState cstate = arg;
+	close_program_pipes(cstate, !IsAbortInProgress());
```
`close_program_pipes` may throw `elog(ERROR`. Is it valid to throw errors during memory context cleanup?

Stolb27 replied:
```
(gdb) bt
#0  close_program_pipes (cstate=0x5568b7c53478, ifThrow=-1 '\377') at copy.c:8014
#1  0x00005568af58d1ce in close_program_pipes_on_reset (arg=0x5568b7c57350) at copy.c:7960
#2  0x00005568afa84955 in MemoryContextCallResetCallbacks (context=0x5568b7c52fe8) at mcxt.c:357
#3  0x00005568afa8477e in MemoryContextDeleteImpl (context=0x5568b7c52fe8, sfile=0x5568b0230880 "mcxt.c",
    func=0x5568b0230ff0 <__func__.7907> "MemoryContextDeleteChildren", sline=291) at mcxt.c:262
#4  0x00005568afa84820 in MemoryContextDeleteChildren (context=0x5568b7bfda20) at mcxt.c:291
#5  0x00005568afa8ac7d in AtAbort_Portals () at portalmem.c:861
#6  0x00005568af40f9b1 in AbortTransaction () at xact.c:3305
#7  0x00005568af41055f in AbortCurrentTransaction () at xact.c:3926
#8  0x00005568af8b038c in PostgresMain (argc=1, argv=0x5568b7b9ea58, dbname=0x5568b7b9e930 "test", username=0x5568b7b9e910 "givi") at postgres.c:5073
#9  0x00005568af80dea0 in BackendRun (port=0x5568b7bc79e0) at postmaster.c:4839
#10 0x00005568af80d458 in BackendStartup (port=0x5568b7bc79e0) at postmaster.c:4503
#11 0x00005568af808cc6 in ServerLoop () at postmaster.c:1953
#12 0x00005568af808070 in PostmasterMain (argc=5, argv=0x5568b7b9c920) at postmaster.c:1523
#13 0x00005568af6eee3e in main (argc=5, argv=0x5568b7b9c920) at main.c:245
```

RekGRpth replied:
I changed it so that the callback would trigger cleanup only on errors.

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1348904

RekGRpth committed:
callback works only on error

RekGRpth committed:
dismiss

InnerLife0 commented:
The reason of lost `close_program_pipes(cstate, false);` is a [merge](https://github.com/arenadata/gpdb/commit/25a90396cd1c52d252a37986e12b63e8e037aa83) with PostgreSQL 9.1beta2.
The [mentioned line](https://github.com/arenadata/gpdb/commit/25a90396cd1c52d252a37986e12b63e8e037aa83#diff-d0556f7c1dc3243be28e24bcc3396b180a949d233006a9e91b4d26930b39164cL2019) was not moved to new function source or to any place. The code of DoCopy() was heavily [changed](https://github.com/arenadata/gpdb/blob/25a90396cd1c52d252a37986e12b63e8e037aa83/src/backend/commands/copy.c#L916) in this patch and line was not moved to new source. Other calls to close_program_pipes() were moved.

So yes, this looks like an accidental mistake.

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1349762

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1349763

RekGRpth committed:
simplify

RekGRpth committed:
indent

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/70101

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1350229

Stolb27 commented on src/backend/commands/copy.c:
```diff
@@ -7955,9 +7955,21 @@ GetTargetSeg(GpDistributionData *distData, Datum *baseValues, bool *baseNulls)
+
+	CopyState	cstate = arg;
+
+	close_program_pipes(cstate, false);
```
Maybe reuse web-table approach from `destroy_execute_handle` to avoid patching pclose_with_stderr?

```suggestion
	if (cstate->program_pipes->pipes[EXEC_DATA_P] != -1)
		close(cstate->program_pipes->pipes[EXEC_DATA_P]);

	/* We don't bother reading possible error message from the pipe */
	if (cstate->program_pipes->pipes[EXEC_ERR_P] != -1)
		close(cstate->program_pipes->pipes[EXEC_ERR_P]);
	if (h->pid != -1)
	{
#ifndef WIN32
		int			status;
		waitpid(cstate->program_pipes->pid, &status, 0);
#endif
	}
```

Stolb27 replied:
Closed as over-optimization.

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1350222

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1350223

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1351361
